### PR TITLE
Improve MongoDB storage module

### PR DESCRIPTION
**Pull request by idegtiarov**
Improved MongoDB usage with using advantages of MongoDB

### CODE REVIEW

File: edx_adapt/data/course_repository.py [hunk 1]
```diff
@@ -321,3 +322,204 @@ def _get_and_assert_problem_exists(self, course_id, problem_dict):
...
+        :param skill_name: name of the added skill
+        """
+        self.store.course_append(course_id, 'skills', skill_name)
+
+    def _add_problem(self, course_id, skill_names, problem_name, tutor_url, b_pretest, b_posttest):
+        """
+        Internal class method for problem fulfilling
+
+        :param course_id: ID of the Course
+        :param skill_names: list of skills name
```
Comment (z4y4ts) [inline edx_adapt/data/course_repository.py]: names

File: edx_adapt/data/course_repository.py [hunk 2]
```diff
@@ -321,3 +322,204 @@ def _get_and_assert_problem_exists(self, course_id, problem_dict):
...
+        :param problem_name: string of problem name
+        :param tutor_url: string with tutor url
+        :param b_pretest: boolean shows pretested option
```
Comment (z4y4ts) [inline edx_adapt/data/course_repository.py]: Do you mean "boolean flag to mark problem as pretest"?

File: edx_adapt/data/course_repository.py [hunk 3]
```diff
@@ -321,3 +322,204 @@ def _get_and_assert_problem_exists(self, course_id, problem_dict):
...
+
+    def get_skills(self, course_id):
+        return self.store.course_get(course_id, 'skills')
+
+    def get_course_ids(self):
+        return self.store.get_tables()
+
+    def get_problems(self, course_id, skill_name=None):
+        """
+        Get all problems related to this course-skill pair:, pretest, normal, and posttest
```
Comment (z4y4ts) [inline edx_adapt/data/course_repository.py]: `:,`

File: edx_adapt/data/mongodb_new_storage.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
+from bson.json_util import dumps
+from json import loads
+
+import pymongo
+
+from edx_adapt import logger
+from . import interface
```
Comment (idegtiarov) [inline edx_adapt/data/mongodb_new_storage.py]: sure, done

File: edx_adapt/data/mongodb_new_storage.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+
+DIRECTION_MAP = {'ascending': pymongo.ASCENDING, 'descending': pymongo.DESCENDING}
+
+
+class MongoDbStorage(interface.StorageInterface):
```
Comment (idegtiarov) [inline edx_adapt/data/mongodb_new_storage.py]: done

File: edx_adapt/data/mongodb_new_storage.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    def set(self, table_name, key, val):
+        self.db[table_name].update_one(
+            {'key': key},
+            {'$set': {'val': val}}, upsert=True)
+
+    def append(self, table_name, list_key, val):
+        table = self.db[table_name]
+        if table.count({'key': list_key}) < 1:
+            raise interface.DataException("List: {0} not in table: {1}".format(list_key, table_name))
+        l = table.find_one({'key': list_key})['val']
```
Comment (idegtiarov) [inline edx_adapt/data/mongodb_new_storage.py]: improved

File: edx_adapt/data/course_repository.py [hunk 7]
```diff
@@ -321,3 +322,204 @@ def _get_and_assert_problem_exists(self, course_id, problem_dict):
...
             # above line will raise exception if it can't be found
         return problem_dict
+
+
+COLL_SUFFIX = {'log': '_log', 'user_problem': '_problems'}
+
+
+class CourseRepositoryMongo(CourseRepository):
+    """
+    Interface specification for MongoDB backend
```
Comment (idegtiarov) [inline edx_adapt/data/course_repository.py]: done

File: edx_adapt/data/course_repository.py [hunk 8]
```diff
@@ -321,3 +322,204 @@ def _get_and_assert_problem_exists(self, course_id, problem_dict):
...
+        """
+        coll_log = course_id + COLL_SUFFIX['log']
+        coll_user_problem = course_id + COLL_SUFFIX['user_problem']
+        self.store.create_table('Courses')
+        self.store.create_table(
+            coll_log, index_fields=[['student_id', 'ascending'], ['timestamp', 'descending'],
+                                    ['problem.problem_name', 'ascending'], ['attempt', 'ascending']],
+            index_unique=True)
+        self.store.create_table(coll_user_problem, index_fields=[['student_id', 'ascending']])
+        data_dict = {'course_id': course_id, 'users_in_progress': [], 'users_finished': [], 'skills': [],
```
Comment (idegtiarov) [inline edx_adapt/data/course_repository.py]: done

File: edx_adapt/data/course_repository.py [hunk 9]
```diff
@@ -321,3 +322,204 @@ def _get_and_assert_problem_exists(self, course_id, problem_dict):
...
+                     'problems': [], 'experiments': []}
+        self.store.record_data(table='Courses', data=data_dict)
+
+    def post_skill(self, course_id, skill_name):
+        """
+        Add skill in courses.skills fild
```
Comment (idegtiarov) [inline edx_adapt/data/course_repository.py]: sure, improve

File: edx_adapt/data/course_repository.py (hunk 1 above)
Comment (idegtiarov) [inline edx_adapt/data/course_repository.py]: done

File: edx_adapt/data/course_repository.py (hunk 2 above)
Comment (idegtiarov) [inline edx_adapt/data/course_repository.py]: seams so %)

File: edx_adapt/data/course_repository.py (hunk 3 above)
Comment (idegtiarov) [inline edx_adapt/data/course_repository.py]: done

File: edx_adapt/data/course_repository.py [hunk 10]
```diff
@@ -321,3 +322,204 @@ def _get_and_assert_problem_exists(self, course_id, problem_dict):
...
+    def get_in_progress_users(self, course_id):
+        return self.store.course_get(course_id, 'users_in_progress')
+
+    def get_finished_users(self, course_id):
+        return self.store.course_get(course_id, 'users_finished')
+
+    def _get_problem(self, course_id, problem_name):
+        problems = self.store.course_search(
+                course_id, 'problems', {'problem_name': problem_name}, 'problems', {'problem_name': problem_name}
+        )['problems']
```
Comment (idegtiarov) [inline edx_adapt/data/course_repository.py]: fix

File: edx_adapt/data/course_repository.py [hunk 11]
```diff
@@ -321,3 +322,204 @@ def _get_and_assert_problem_exists(self, course_id, problem_dict):
...
+        data = {'student_id': user_id, 'problem': problem, 'unix_s': unix_seconds, 'type': 'page_load',
+                'timestamp': datetime.datetime.fromtimestamp(unix_seconds).strftime('%Y-%m-%d %H:%M:%S')}
+        # TODO(idegtiarov) check that load was already stored for this problem and do not log. But if we interested is
+        # statistic it seems logging additional/same data is ok
+        self.store.record_data(coll, data)
+
+    def set_next_problem(self, course_id, user_id, problem_dict):
+        # TODO(idegtiarov) check that problem described in problem_dict exists in generic problem list
+        coll = course_id + COLL_SUFFIX['user_problem']
+        # current = self.get_next_problem(course_id, user_id)
```
Comment (idegtiarov) [inline edx_adapt/data/course_repository.py]: done

File: edx_adapt/data/course_repository.py [hunk 12]
```diff
@@ -407,41 +424,71 @@ def get_finished_users(self, course_id):
...
-        data = {'student_id': user_id, 'problem': problem, 'correct': correct, 'attempt': attempt,
-                'unix_s': unix_seconds, 'type': 'response',
-                'timestamp': datetime.datetime.fromtimestamp(unix_seconds).strftime('%Y-%m-%d %H:%M:%S')}
+        data = {
+            'student_id': user_id,
+            'problem': problem,
+            'correct': correct,
+            'attempt': attempt,
+            'unix_s': unix_seconds, 'type': 'response',
+            'timestamp': datetime.datetime.fromtimestamp(unix_seconds).strftime('%Y-%m-%d %H:%M:%S')
```
Comment (z4y4ts) [inline edx_adapt/data/course_repository.py]: I usually do `from datetime import datetime` to avoid calling `datetime.datetime....`

Comment (idegtiarov) [inline edx_adapt/data/course_repository.py]: Make sense will improve and merge

**Pull request merged**